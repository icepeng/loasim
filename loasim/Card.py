from typing import List, Tuple
from loasim.CharacterModifier import CharacterModifier


class Card:
    def __init__(self, name: str, mdf: CharacterModifier) -> None:
        self.name = name
        self.mdf = mdf

    def get_modifier(self):
        return self.mdf


card_list = [
    Card("남겨진 바람의 절벽 (12)", CharacterModifier(crit=7)),
    Card("세상을 구하는 빛 (18)", CharacterModifier(pdamage_indep=7)),
    Card("세상을 구하는 빛 (30)", CharacterModifier(pdamage_indep=15)),
]

card_dict = {card.name: card for card in card_list}


class CardManager:
    def __init__(self, *equipped_list: List[str]) -> None:
        self.equipped_list = equipped_list

    def get_static_modifier(self):
        mdf = CharacterModifier()
        for name in self.equipped_list:
            card = card_dict.get(name)
            if card is None:
                raise Exception(f"{name} is not an card")
            mdf = mdf + card.get_modifier()
        return mdf
