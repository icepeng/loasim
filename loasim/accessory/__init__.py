from loasim.accessory.builtin import lostark_accessory_repository
from loasim.accessory.base import AccessoryState
