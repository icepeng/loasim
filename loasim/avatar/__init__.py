from loasim.avatar.base import AvatarState
from loasim.avatar.builtin import lostark_avatar_repository
