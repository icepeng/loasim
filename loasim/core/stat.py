from __future__ import annotations

from math import sqrt

from pydantic import BaseModel, Extra


class InternalStat(BaseModel):
    class Config:
        extra = Extra.forbid

    weapon_att: float = 0
    stat_main: float = 0
    pstat_main: float = 0
    crit: float = 0
    special: float = 0
    swift: float = 0

    def get_stat(self) -> Stat:
        return Stat(
            crit=self.crit * 0.03577,
            crit_damage=200,
            att=sqrt(
                self.weapon_att * (self.stat_main * (100 + self.pstat_main) / 100) / 6
            ),
            moving_speed=self.swift * 0.01717,
        )

    def __add__(self, arg: InternalStat) -> InternalStat:
        return InternalStat(
            weapon_att=self.weapon_att + arg.weapon_att,
            stat_main=self.stat_main + arg.stat_main,
            pstat_main=self.pstat_main + arg.pstat_main,
            crit=self.crit + arg.crit,
            special=self.special + arg.special,
            swift=self.swift + arg.swift,
        )


def add_pdamage_indep(a: float, b: float) -> float:
    return a + b + (a * b) * 0.01


def sub_pdamage_indep(a: float, b: float) -> float:
    return (100 + a) / (100 + b) * 100 - 100


class Stat(BaseModel):
    class Config:
        extra = Extra.forbid

    crit: float = 0
    crit_damage: float = 0
    crit_damage_head: float = 0
    crit_damage_back: float = 0
    pdamage: float = 0
    pdamage_indep: float = 0
    pdamage_indep_head: float = 0
    pdamage_indep_back: float = 0
    armor_ignore: float = 0
    patt_common: float = 0
    patt_job: float = 0
    patt_indep: float = 0
    att: float = 0
    moving_speed: float = 0

    def __add__(self, arg: Stat) -> Stat:
        return Stat(
            crit=self.crit + arg.crit,
            crit_damage=self.crit_damage + arg.crit_damage,
            crit_damage_head=self.crit_damage_head + arg.crit_damage_head,
            crit_damage_back=self.crit_damage_back + arg.crit_damage_back,
            pdamage=self.pdamage + arg.pdamage,
            pdamage_indep=add_pdamage_indep(self.pdamage_indep, arg.pdamage_indep),
            pdamage_indep_head=add_pdamage_indep(
                self.pdamage_indep_head, arg.pdamage_indep_head
            ),
            pdamage_indep_back=add_pdamage_indep(
                self.pdamage_indep_back, arg.pdamage_indep_back
            ),
            armor_ignore=100
            - 0.01 * ((100 - self.armor_ignore) * (100 - arg.armor_ignore)),
            patt_common=self.patt_common + arg.patt_common,
            patt_job=self.patt_job + arg.patt_job,
            patt_indep=add_pdamage_indep(self.patt_indep, arg.patt_indep),
            att=self.att + arg.att,
            moving_speed=self.moving_speed + arg.moving_speed,
        )

    def __sub__(self, arg: Stat) -> Stat:
        return Stat(
            crit=self.crit - arg.crit,
            crit_damage=self.crit_damage - arg.crit_damage,
            crit_damage_head=self.crit_damage_head - arg.crit_damage_head,
            crit_damage_back=self.crit_damage_back - arg.crit_damage_back,
            pdamage=self.pdamage - arg.pdamage,
            pdamage_indep=sub_pdamage_indep(self.pdamage_indep, arg.pdamage_indep),
            pdamage_indep_head=sub_pdamage_indep(
                self.pdamage_indep_head, arg.pdamage_indep_head
            ),
            pdamage_indep_back=sub_pdamage_indep(
                self.pdamage_indep_back, arg.pdamage_indep_back
            ),
            armor_ignore=100
            - 100 * (100 - self.armor_ignore) / (100 - arg.armor_ignore),
            patt_common=self.patt_common - arg.patt_common,
            patt_job=self.patt_job - arg.patt_job,
            patt_indep=sub_pdamage_indep(self.patt_indep, arg.patt_indep),
            att=self.att - arg.att,
            moving_speed=self.moving_speed - arg.moving_speed,
        )

    def get_total_att(self) -> float:
        patt = add_pdamage_indep(
            add_pdamage_indep(self.patt_common, self.patt_job), self.patt_indep
        )
        return self.att * (1 + patt / 100)

    def get_multiplier(self) -> float:
        return (1 + self.pdamage_indep / 100) * (1 + self.pdamage / 100)
