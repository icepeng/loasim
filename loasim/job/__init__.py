from loasim.job.artist import lostark_job_artist
from loasim.job.base import Job, SkillSpecification, SkillState, Tripod
from loasim.job.blade import lostark_job_blade
from loasim.job.hawkeye import lostark_job_hawkeye
from loasim.job.scouter import lostark_job_scouter

job_dict = {
    "artist": lostark_job_artist,
    "blade": lostark_job_blade,
    "hawkeye": lostark_job_hawkeye,
    "scouter": lostark_job_scouter,
}


def get_job(job_name: str) -> Job:
    job = job_dict[job_name]
    if job is None:
        raise ValueError(f"{job_name} is not supported")
    return job
