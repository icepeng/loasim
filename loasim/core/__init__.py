from loasim.core.buff import (
    BuffManager,
    OnoffBuff,
    SkillBuff,
    StackBuff,
    StatBuff,
    StaticBuff,
)
from loasim.core.enemy import Enemy
from loasim.core.skill import Skill, SkillSpecification, Tripod
from loasim.core.stat import InternalStat, Stat
