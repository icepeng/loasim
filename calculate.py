from collections import defaultdict
from typing import List
from loasim.core import Skill


def calculate(dealcycle: List[Skill], cycle_time, enemy):
    total_damage = 0
    damage_dict = defaultdict(float)
    for sk in dealcycle:
        damage = sk.get_damage(enemy)
        total_damage += damage
        damage_dict[sk.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(f"dps: {total_damage / cycle_time:,.0f}")
