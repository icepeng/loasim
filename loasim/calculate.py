from loguru import logger

from collections import defaultdict
from typing import Any, Callable, List, Dict, Literal, Tuple

from pydantic import BaseModel

from loasim.core import Skill, Enemy, BuffManager, InternalStat
from loasim.core.buff import Buff
from loasim.core.stat import Stat
from loasim.setitem import lostark_setitem_repository
from loasim.card import lostark_default_card_repository
from loasim.engraving import lostark_engraving_repository


class DealCycle(BaseModel):
    buff_list: List[Buff]
    skill_list: List[Tuple[Skill, Dict[str, Any], Literal["back", "head"] | None]]
    cycle_time: float


def calculate(
    internal_stat: InternalStat,
    weapon_pdamage: float,
    setitem_state: List[Tuple[str, int, int]],
    engraving_state: List[Tuple[str, int]],
    card_state: List[str],
    generator: Callable[[InternalStat], DealCycle],
    enemy: Enemy,
):
    deal_cycle = generator(internal_stat)
    base_stat = internal_stat.get_stat()
    weapon_stat = Stat(pdamage=weapon_pdamage)  # 무기 품질
    basis_stat = (
        base_stat
        + weapon_stat
        + lostark_setitem_repository.get_stat(setitem_state)
        + lostark_default_card_repository.get_stat(card_state)
    )

    logger.info(base_stat)
    logger.info(basis_stat)

    engraving_buffs = lostark_engraving_repository.get_buffs(engraving_state)
    buff_manager = BuffManager(engraving_buffs + deal_cycle.buff_list)

    total_damage = 0.0
    damage_dict: defaultdict[str, float] = defaultdict(float)
    for skill, buff_state, backhead in deal_cycle.skill_list:
        buff_stat = buff_manager.get_stat(buff_state, skill)
        damage = skill.get_damage(enemy, backhead, basis_stat + buff_stat)
        total_damage += damage
        damage_dict[skill.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(deal_cycle.cycle_time)
    print(f"dps: {total_damage / deal_cycle.cycle_time:,.0f}")
