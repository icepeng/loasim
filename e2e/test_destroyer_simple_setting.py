from loasim.calculate import DealCycle, calculate
from loasim.core import BuffState, Enemy, InternalStat
from loasim.core.stat import Stat
from loasim.job import SkillState


def calculate_destroyer():
    core_indure_state = {
        "고통의 흔적": BuffState(onoff=True),
        "분노의 망치": BuffState(onoff=False),
        "중력 코어": BuffState(onoff=False),
    }
    core_no_indure_state = {
        "고통의 흔적": BuffState(onoff=False),
        "분노의 망치": BuffState(onoff=False),
        "중력 코어": BuffState(onoff=False),
    }
    release_indure_state = {
        "고통의 흔적": BuffState(onoff=True),
        "분노의 망치": BuffState(onoff=True),
        "중력 코어": BuffState(onoff=True, stack=3),
    }
    release_no_indure_state = {
        "고통의 흔적": BuffState(onoff=False),
        "분노의 망치": BuffState(onoff=True),
        "중력 코어": BuffState(onoff=True, stack=3),
    }

    deal_cycle: DealCycle = [
        ("사이즈믹 해머", release_indure_state, "head"),
        ("인듀어 페인", core_no_indure_state, "head"),
        ("그라비티 임팩트", core_indure_state, "head"),
        ("풀 스윙", release_indure_state, "head"),
        ("헤비 크러쉬", core_indure_state, "head"),
        ("러닝 크래쉬", core_no_indure_state, "head"),
        ("퍼펙트 스윙", release_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("그라비티 임팩트", core_no_indure_state, "head"),
        ("풀 스윙", release_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("드레드노트", core_no_indure_state, "head"),
        ("사이즈믹 해머", release_no_indure_state, "head"),
        ("그라비티 임팩트", core_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("풀 스윙", release_no_indure_state, "head"),
        ("러닝 크래쉬", core_no_indure_state, "head"),
        ("인듀어 페인", core_no_indure_state, "head"),
        ("퍼펙트 스윙", release_indure_state, "head"),
        ("헤비 크러쉬", core_indure_state, "head"),
        ("그라비티 임팩트", core_no_indure_state, "head"),
        ("풀 스윙", release_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("드레드노트", core_no_indure_state, "head"),
        ("사이즈믹 해머", release_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("그라비티 임팩트", core_no_indure_state, "head"),
        ("풀 스윙", release_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("러닝 크래쉬", core_no_indure_state, "head"),
        ("퍼펙트 스윙", release_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("그라비티 임팩트", core_no_indure_state, "head"),
        ("풀 스윙", release_no_indure_state, "head"),
        ("헤비 크러쉬", core_no_indure_state, "head"),
        ("드레드노트", core_no_indure_state, "head"),
    ]

    return calculate(
        job_name="destroyer",
        internal_stat=InternalStat(
            weapon_att=52051, stat_main=198850, crit=1471, special=449, swift=544
        ),
        weapon_pdamage=28.44,
        setitem_state=[("사멸", 6, 6)],
        engraving_state=[
            ("원한", 3),
            ("슈퍼 차지", 3),
            ("분노의 망치", 3),
            ("바리케이드", 3),
            ("저주받은 인형", 3),
        ],
        card_state=["세상을 구하는 빛 (18)"],
        skill_state={
            "헤비 크러쉬": SkillState(
                level=12,
                gem=0,
                tripod={},
            ),
            "인듀어 페인": SkillState(
                level=10,
                gem=0,
                tripod={
                    "고통의 흔적": 1,
                },
            ),
            "그라비티 임팩트": SkillState(level=10, gem=0, tripod={}),
            "드레드노트": SkillState(
                level=11,
                gem=0,
                tripod={
                    "몰아치는 해머": 5,
                },
            ),
            "러닝 크래쉬": SkillState(
                level=7,
                gem=0,
                tripod={},
            ),
            "퍼펙트 스윙": SkillState(
                level=12,
                gem=9,
                tripod={
                    "약점포착": 5,
                    "날카로운 해머": 5,
                    "무절제": 5,
                },
            ),
            "사이즈믹 해머": SkillState(
                level=12,
                gem=9,
                tripod={
                    "절대적인 힘": 5,
                    "날카로운 벽": 5,
                    "굶주린 힘": 5,
                },
            ),
            "풀 스윙": SkillState(
                level=12,
                gem=9,
                tripod={
                    "빠른 준비": 5,
                    "무서운 해머": 5,
                    "야수의 눈": 5,
                },
            ),
        },
        enemy=Enemy(armor=6000, reduction=23),
        deal_cycle=deal_cycle,
        cycle_time=68,
        additional_stat=Stat(
            pdamage_indep=3 / 3, crit=5 / 3, crit_damage=8 / 3
        ),  # 순환 팔찌
    )


def test_destroyer_simple_setting(snapshot: float):
    assert calculate_destroyer() == snapshot


if __name__ == "__main__":
    calculate_destroyer()
