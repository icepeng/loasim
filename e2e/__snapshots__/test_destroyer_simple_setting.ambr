# name: test_destroyer_simple_setting
  7647357.869114211
---
